index.bs

```diff
@@ -2725,6 +2773,249 @@ The [=remote end steps=] with <var ignore>session</var> and |command parameters|
 
 1. Return [=success=] with data null.
 
+</div>
+#### The browsingContext.locateNodes Command ####  {#command-browsingContext-locateNodes}
+
+The <dfn export for=commands>browsingContext.locateNodes</dfn> command returns a
+list of all nodes matching the specified locators.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      browsingContext.LocateNodes = (
+        method: "browsingContext.locateNodes",
+        params: browsingContext.LocateNodesParameters
+      )
+
+      browsingContext.LocateNodesParameters = {
+         context: browsingContext.BrowsingContext,
+         locator: browsingContext.Locator,
+         startNodes: [ * script.SharedReference ],
+         ? maxNodes: js-uint,
+      }
+      </pre>
+   </dd>
+   <dt>Result Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+        browsingContext.LocateNodesResult = {
+            context: browsingContext.BrowsingContext,
+            nodeSharedIds: [ * script.SharedId ]
+        }
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="locate nodes using CSS">
+To <dfn>locate nodes using CSS</dfn> with given |context target|, |context nodes|,
+|selector| and <var ignore>session</var>:
+
+1. Let |returned nodes| be an empty array.
+
+1. For each |context node| of |context nodes|:
+
+  1. Let |parse result| be the result of [=parse a selector=] given |selector|.
+
+  1. If |parse result| is failure, return [=error=] with [=error code=] [=invalid selector=].
+
+  1. Let |elements| be the result of [=match a selector against a tree=] with 
+     |parse result| and |context target|’s [=active document=] [=root=] using
+     [=scoping root=] |context node|.
+
+  1. For each |element| in |elements|:
+
+    1. Append |element| to |returned nodes|
+
+1. Return |returned nodes|.
+
+</div>
+
+<div algorithm="locate nodes using XPath">
+To <dfn>locate nodes using XPath</dfn> with given |context target|,
+|context nodes|, |selector| and <var ignore>session</var>:
+
+1. Let |returned nodes| be an empty array.
+
+1. For each |context node| of |context nodes|:
+
+   1. Let |evaluateResult| be the result of calling [=evaluate=] on |context target|'s
+      [=active document=], with arguments |selector|, |context node|, null,
+      [=ORDERED_NODE_SNAPSHOT_TYPE=], and null.
+
+   1. Let |index| be 0.
+
+   1. Let |length| be the result of getting the <code>snapshotLength</code> property
+      from |evaluateResult|. If this throws an [=XPathException=] return [=error=]
+      with [=error code=] [=invalid selector=], otherwise if this throws any other
+      exception return return [=error=] with [=error code=] [=unknown error=].
+
+   1. Repeat, while |index| is less than |length|:
+
+      1. Let |node| be the result of calling [=snapshotItem=] with |evaluateResult|
+         as this and |index| as the argument.
+
+      1. If |node| is not an element return [=error=] with [=error code=]
```

Comment:
It definitely seems like we could support returning non-Element nodes here. Maybe even non-nodes if we can select them.

Reply:
@jgraham We can support returning non-element nodes, and I'll update the PR to reflect that. However, I would like to retain the restriction to only return nodes from XPath, since the command is `locateNodes`.